### PR TITLE
Python: Action Planner implementation

### PR DESCRIPTION
**Pull request by Sailesh-2209**
### Motivation and Context

This PR is a first draft for implementation of ActionPlanner in Python. I try to follow the C# implementation as much as possible.
Fixes: #1917

### Description

- I have implemented ActionPlanner in the `python/semantic_kernel/planning/action_planner` directory. The implementation is in a file called `action_planner.py` and the prompt used by the planner (a replica of the prompt used in C#) is found in `skprompt.txt` in the same directory.
- The functions `good_examples`, `edge_case_examples` and `list_of_functions` are used by the prompt to populate examples and list of available functions in the kernel.
- After creating the plan, I parse the JSON generated by the planner by using regex matching to extract out pure JSON as implemented in #1496 
- Based on whether or not the planner was able to identify a function, I create a Plan instance and return it.

### Here is some example code for using the planner
```python
import logging
import semantic_kernel as sk
from semantic_kernel.planning import ActionPlanner
from semantic_kernel.connectors.ai.open_ai import (
	OpenAIChatCompletion,
	OpenAITextCompletion
)
from semantic_kernel.core_skills import (
	MathSkill,
	FileIOSkill,
	TimeSkill,
	TextSkill
)

async def main():
	kernel = sk.Kernel()
	api_key, org_id = sk.openai_settings_from_dot_env()

	kernel.add_text_completion_service("dv", OpenAITextCompletion("text-davinci-003", api_key, org_id))
	kernel.import_skill(MathSkill(), "math")
	kernel.import_skill(FileIOSkill(), "fileIO")
	kernel.import_skill(TimeSkill(), "time")
	kernel.import_skill(TextSkill(), "text")

	planner = ActionPlanner(kernel)
	ask = "What is the sum of 110 and 990?"
	plan = await planner.create_plan_async(goal=ask)
	result = await plan.invoke_async()
	print(result.result)

if __name__ == "__main__":
	import asyncio
	asyncio.run(main())
```

### Queries
- In the C# implementation, when populating the function parameters in the plan object, the `plan.parameters` variable is updated with the function parameters. When I did the same in python, these parameters weren't being passed onto the function in `invoke_async` method of the plan instance. The parameters are only being passed when I update them in the `plan.state` variable. This code is in `line 121` of `action_planner.py`. What is the right way to do this?
- Some skills / plugins require one of the inputs to be passed as the `input` variable and not in the context. How should this be handled in the planner? 
image
**OR**
image

### Issues that I faced
- The `text-davinci-002` text-completion model regularly output incorrectly formatted JSON with commas missing. For example
  ```json
  {""plan"":{
  ""rationale"": ""The list contains a function that allows for writing to files, so the plan is to use that function to write the desired content to the file 'output.txt'."",
  ""function"": ""fileIO.writeAsync"",
  ""parameters"": {
  ""input"": null,
  ""content"": ""A brief history of computer science""
  ""path"": ""output.txt""
  }}}
  ```

Pending Items
- [x] Create unit tests
- [x] Send the function parameters to the planner in list of available functions with/without input parameter.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by alexchaomander** [comment]
With https://github.com/microsoft/semantic-kernel/pull/1907 merged in, we'll make sure to review this in detail this week!

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
@Sailesh-2209 I am doing some local testing and will get back to you soon.

While I'm doing this, can you add an example usage file to `python/samples/kernel-syntax-examples` and remove the changes to the launch.json file? Thanks!

**Comment by Sailesh-2209** [comment]
@awharrison-28 
- Removed changes to launch.json
- Added example to `python/samples/kernel-syntax-examples`
- I was not passing arguments correctly while creating a Plan object, fixed that.
- the new line at the end of `python/semantic_kernel/core_skills/text_skill.py` is from running `pre-commit` 😅

File: python/semantic_kernel/planning/action_planner/action_planner.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+        generated_plan_raw = await self._planner_function.invoke_async(
+            context=self._context
+        )
+        generated_plan_raw_str = str(generated_plan_raw)
+
+        self._logger.info(f"Plan generated by ActionPlanner:\n{generated_plan_raw_str}")
+
+        # Ignore additional text around JSON recursively
+        json_regex = r"\{(?:[^{}]|(?R))*\}"
```
Comment (awharrison-28) [inline python/semantic_kernel/planning/action_planner/action_planner.py:85]: There are a number of scenarios where AI will not generate a valid plan: perhaps it doesn't think the current list of functions are appropriate, service throttling, etc.

I think there should be a check here on whether a plan was actually found.

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/samples/kernel-syntax-examples/action_planner.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import semantic_kernel as sk
```
Comment (awharrison-28) [inline python/samples/kernel-syntax-examples/action_planner.py:1]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

import semantic_kernel as sk
```

File: python/semantic_kernel/planning/__init__.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
+from semantic_kernel.planning.action_planner.action_planner import ActionPlanner
```
Comment (awharrison-28) [inline python/semantic_kernel/planning/__init__.py:1]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

from semantic_kernel.planning.action_planner.action_planner import ActionPlanner
```

File: python/semantic_kernel/planning/action_planner/action_planner.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
+import json
```
Comment (awharrison-28) [inline python/semantic_kernel/planning/action_planner/action_planner.py:1]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

import json
```

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/planning/action_planner/action_planner.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        self._logger.info(f"Finding the best function for achieving the goal: {goal}")
+
+        self._context.variables.update(goal)
+
+        generated_plan_raw = await self._planner_function.invoke_async(
+            context=self._context
+        )
+        generated_plan_raw_str = str(generated_plan_raw)
+
+        if not generated_plan_raw or not generated_plan_raw_str:
```
Comment (awharrison-28) [inline python/semantic_kernel/planning/action_planner/action_planner.py:84]: This doesn't account for a malformed plan, suggestion below:

Suggestion: 

```
        generated_plan_raw_str = str(generated_plan_raw)

        self._logger.info(f"Plan generated by ActionPlanner:\n{generated_plan_raw_str}")

        # Ignore additional text around JSON recursively
        json_regex = r"\{(?:[^{}]|(?R))*\}"
        generated_plan_str = regex.search(json_regex, generated_plan_raw_str)

        if generated_plan_str is None or generated_plan_str == "":
            self._logger.error("No valid plan has been generated.")
            return Plan(description=goal)
        
        generated_plan_str = generated_plan_str.group().replace('""', '"')
```

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

**Comment by awharrison-28** [comment]
hi @Sailesh-2209 I added the `planning_exception.py` file from https://github.com/microsoft/semantic-kernel/pull/1916 to this for consistency and raised these exceptions where errors were logged or ValueExceptions raised. Could you take a look and see if you agree? I think it'd be good to add some basic unit tests to ensure that the methods run as expected (don't worry about integration tests, the syntax example is good for now).

**Comment by Sailesh-2209** [comment]
The exceptions LGTM! I didn't find the time to add the tests today, will do it tomorrow.

**Comment by awharrison-28** [comment]
@Sailesh-2209 https://github.com/microsoft/semantic-kernel/pull/1916 went in today, so you can likely reuse some of the unit test structures and fakes for sequential planner :)

**Comment by Sailesh-2209** [comment]
@awharrison-28 I've started writing the unit tests. I will need some more time to finish it up. Hopefully it will be done by tomorrow.

**Review by awharrison-28 — APPROVED** [review]

**Pull request merged**